File: Other Versions/Go/notifications/notifications.go
// Package notifications sends cross-platform OS-native desktop notifications.
//
// Dispatches to the right backend based on detected platform:
//   - Linux / WSL: `notify-send` (from libnotify)
//   - macOS:       `osascript`
//   - Windows:     PowerShell using the built-in Windows.UI.Notifications APIs
//
// All backends shell out via os/exec — no external Go dependencies.
// If the OS backend is unavailable (e.g., no notify-send installed), a warning
// is logged on first use and all subsequent calls return false without error.
package notifications

import (
	"context"
	"log"
	"os/exec"
	"strings"
	"sync"
	"time"

	"aihound/core"
)

// Urgency levels. These are Linux notify-send values; other backends map them
// as best they can.
const (
	UrgencyLow      = "low"
	UrgencyNormal   = "normal"
	UrgencyCritical = "critical"
)

var (
	capabilityOnce      sync.Once
	capabilityAvailable bool

	// Logger receives warnings about unavailable backends. Users who want to
	// silence logs can assign a discard logger.
	Logger = log.Default()
)

// checkCapability tests whether the current platform can send notifications.
// Cached after first call.
func checkCapability() bool {
	capabilityOnce.Do(func() {
		plat := core.DetectPlatform()
		switch plat {
		case core.PlatformLinux, core.PlatformWSL:
			if _, err := exec.LookPath("notify-send"); err == nil {
				capabilityAvailable = true
			} else {
				Logger.Print("desktop notifications unavailable: `notify-send` not found. " +
					"Install libnotify-bin (apt) / libnotify (dnf) / equivalent to enable.")
			}
		case core.PlatformMacOS:
			if _, err := exec.LookPath("osascript"); err == nil {
				capabilityAvailable = true
			} else {
				Logger.Print("desktop notifications unavailable: `osascript` not found.")
			}
		case core.PlatformWindows:
			if _, err := exec.LookPath("powershell.exe"); err == nil {
				capabilityAvailable = true
			} else if _, err := exec.LookPath("powershell"); err == nil {
				capabilityAvailable = true
			} else {
				Logger.Print("desktop notifications unavailable: PowerShell not found.")
			}
		default:
			capabilityAvailable = false
		}
	})
	return capabilityAvailable
}

// SendNotification sends a desktop notification. Returns true on success.
//
// Non-fatal: if the backend is unavailable or the command fails, returns false
// without panicking. Callers may ignore the return value.
func SendNotification(title, body, urgency string) bool {
	if !checkCapability() {
		return false
	}

	defer func() {
		// Defensive: never propagate a panic out of the notification path.
		_ = recover()
	}()

	plat := core.DetectPlatform()
	switch plat {
	case core.PlatformLinux, core.PlatformWSL:
		return notifyLinux(title, body, urgency)
	case core.PlatformMacOS:
		return notifyMacOS(title, body, urgency)
	case core.PlatformWindows:
		return notifyWindows(title, body, urgency)
	}
	return false
}

// runWithTimeout runs a command with a hard timeout and returns whether it
// succeeded. stderr is captured and logged at debug on failure.
func runWithTimeout(timeout time.Duration, name string, args ...string) bool {
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()
	cmd := exec.CommandContext(ctx, name, args...)
	out, err := cmd.CombinedOutput()
	if err != nil {
		Logger.Printf("notification backend %s failed: %v (%s)", name, err, strings.TrimSpace(string(out)))
		return false
	}
	return true
}

func notifyLinux(title, body, urgency string) bool {
	if urgency == "" {
		urgency = UrgencyNormal
	}
	return runWithTimeout(
		5*time.Second,
		"notify-send",
		"--urgency="+urgency,
		"--app-name=AIHound",
		title,
		body,
	)
}

// applescriptSafe escapes a string for safe inclusion in AppleScript by
// splitting on double-quotes and rejoining with the AppleScript `quote` constant.
func applescriptSafe(s string) string {
	parts := strings.Split(s, `"`)
	quoted := make([]string, len(parts))
	for i, p := range parts {
		quoted[i] = `"` + p + `"`
	}
	return strings.Join(quoted, ` & quote & `)
}

func notifyMacOS(title, body, urgency string) bool {
	safeTitle := applescriptSafe(title)
	safeBody := applescriptSafe(body)
	script := `display notification ` + safeBody + ` with title "AIHound" subtitle ` + safeTitle
	if urgency == UrgencyCritical {
		script += ` sound name "Basso"`
	}
	return runWithTimeout(5*time.Second, "osascript", "-e", script)
}

// xmlEscape escapes characters that are special in XML content.
func xmlEscape(s string) string {
	s = strings.ReplaceAll(s, "&", "&amp;")
	s = strings.ReplaceAll(s, "<", "&lt;")
	s = strings.ReplaceAll(s, ">", "&gt;")
	return s
}

func notifyWindows(title, body, urgency string) bool {
	_ = urgency // Windows toast XML here doesn't plumb urgency; retained for API symmetry.

	// XML-escape title and body for safe insertion into the toast template.
	safeTitle := xmlEscape(title)
	safeBody := xmlEscape(body)

	// Build the PowerShell script using a single-quoted here-string (@'...'@)
	// which is non-expanding, then use -replace to inject the escaped values.
	var b strings.Builder
	b.WriteString("$ErrorActionPreference = 'Stop'\n")
	b.WriteString("[void][Windows.UI.Notifications.ToastNotificationManager, Windows.UI.Notifications, ContentType=WindowsRuntime]\n")
	b.WriteString("$template = @'\n")
	b.WriteString("<toast>\n")
	b.WriteString("  <visual>\n")
	b.WriteString("    <binding template=\"ToastGeneric\">\n")
	b.WriteString("      <text>TITLE_PLACEHOLDER</text>\n")
	b.WriteString("      <text>BODY_PLACEHOLDER</text>\n")
	b.WriteString("    </binding>\n")
	b.WriteString("  </visual>\n")
	b.WriteString("</toast>\n")
	b.WriteString("'@\n")
	// Use PowerShell -replace to substitute placeholders with the escaped values.
	// Single-quote the replacement strings and double any embedded single quotes.
	psSafeTitle := strings.ReplaceAll(safeTitle, "'", "''")
	psSafeBody := strings.ReplaceAll(safeBody, "'", "''")
	b.WriteString("$template = $template -replace 'TITLE_PLACEHOLDER', '" + psSafeTitle + "'\n")
	b.WriteString("$template = $template -replace 'BODY_PLACEHOLDER', '" + psSafeBody + "'\n")
	b.WriteString("$xml = [Windows.Data.Xml.Dom.XmlDocument]::new()\n")
	b.WriteString("$xml.LoadXml($template)\n")
	b.WriteString("$toast = [Windows.UI.Notifications.ToastNotification]::new($xml)\n")
	b.WriteString("$notifier = [Windows.UI.Notifications.ToastNotificationManager]::CreateToastNotifier('AIHound')\n")
	b.WriteString("$notifier.Show($toast)\n")
	script := b.String()

	exeName := "powershell.exe"
	if _, err := exec.LookPath(exeName); err != nil {
		if _, err := exec.LookPath("powershell"); err == nil {
			exeName = "powershell"
		}
	}

	return runWithTimeout(
		10*time.Second,
		exeName,
		"-NoProfile",
		"-NonInteractive",
		"-Command",
		script,
	)
}

File: aihound/notifications.py
"""Cross-platform OS-native desktop notifications.

Dispatches to the right backend based on detected platform:
- Linux / WSL (if D-Bus reachable): `notify-send`
- macOS: `osascript`
- Windows: PowerShell (BurntToast if installed, else built-in toast XML)

All backends shell out via stdlib `subprocess`. No new Python dependencies.
If the OS backend is unavailable (e.g., no `notify-send` installed), we log a
warning on first use and all subsequent calls are no-ops.
"""

from __future__ import annotations

import logging
import shutil
import subprocess
import xml.sax.saxutils
from typing import Optional

from aihound.core.platform import detect_platform, Platform

logger = logging.getLogger("aihound.notifications")

# Urgency levels — Linux notify-send values; other backends map these as best they can
URGENCY_LOW = "low"
URGENCY_NORMAL = "normal"
URGENCY_CRITICAL = "critical"

_capability_checked = False
_capability_available = False


def _check_capability() -> bool:
    """Test whether the current platform can actually send notifications.

    Cached after first call.
    """
    global _capability_checked, _capability_available
    if _capability_checked:
        return _capability_available
    _capability_checked = True

    plat = detect_platform()
    if plat == Platform.LINUX or plat == Platform.WSL:
        _capability_available = shutil.which("notify-send") is not None
        if not _capability_available:
            logger.warning(
                "Desktop notifications unavailable: `notify-send` not found. "
                "Install libnotify-bin (apt) / libnotify (dnf) / equivalent to enable."
            )
    elif plat == Platform.MACOS:
        _capability_available = shutil.which("osascript") is not None
        if not _capability_available:
            logger.warning("Desktop notifications unavailable: `osascript` not found.")
    elif plat == Platform.WINDOWS:
        _capability_available = shutil.which("powershell.exe") is not None or shutil.which("powershell") is not None
        if not _capability_available:
            logger.warning("Desktop notifications unavailable: PowerShell not found.")
    else:
        _capability_available = False

    return _capability_available


def send_notification(
    title: str,
    body: str,
    urgency: str = URGENCY_NORMAL,
    icon: Optional[str] = None,
) -> bool:
    """Send a desktop notification. Returns True on success, False on failure.

    Non-fatal — if the backend is unavailable or the command fails, logs a debug
    message and returns False. Never raises.
    """
    if not _check_capability():
        return False

    plat = detect_platform()
    try:
        if plat == Platform.LINUX or plat == Platform.WSL:
            return _notify_linux(title, body, urgency, icon)
        if plat == Platform.MACOS:
            return _notify_macos(title, body, urgency)
        if plat == Platform.WINDOWS:
            return _notify_windows(title, body, urgency)
    except Exception as e:
        logger.debug("Notification backend raised: %s", e)
        return False

    return False


def _notify_linux(title: str, body: str, urgency: str, icon: Optional[str]) -> bool:
    cmd = ["notify-send", "--urgency=" + urgency, "--app-name=AIHound"]
    if icon:
        cmd.append(f"--icon={icon}")
    cmd.extend([title, body])
    result = subprocess.run(cmd, capture_output=True, text=True, timeout=5)
    if result.returncode != 0:
        logger.debug("notify-send failed (rc=%d): %s", result.returncode, result.stderr)
        return False
    return True


def _applescript_safe(s: str) -> str:
    """Encode a Python string as a safe AppleScript expression."""
    parts = s.split('"')
    escaped_parts = [f'"{p}"' for p in parts]
    return " & quote & ".join(escaped_parts)


def _notify_macos(title: str, body: str, urgency: str) -> bool:
    # osascript: display notification <body> with title "AIHound" subtitle <title>
    # Use AppleScript quote constant concatenation for safe escaping
    safe_title = _applescript_safe(title)
    safe_body = _applescript_safe(body)
    script = (
        f'display notification {safe_body} with title "AIHound" '
        f'subtitle {safe_title}'
    )
    # Critical urgency: play a sound
    if urgency == URGENCY_CRITICAL:
        script += ' sound name "Basso"'
    result = subprocess.run(
        ["osascript", "-e", script], capture_output=True, text=True, timeout=5
    )
    if result.returncode != 0:
        logger.debug("osascript failed (rc=%d): %s", result.returncode, result.stderr)
        return False
    return True


def _notify_windows(title: str, body: str, urgency: str) -> bool:
    """Show a Windows toast via PowerShell.

    Uses Windows.UI.Notifications APIs that ship with Windows 10+. No external
    PowerShell modules required.
    """
    # XML-escape to prevent injection into the toast template
    safe_title = xml.sax.saxutils.escape(title)
    safe_body = xml.sax.saxutils.escape(body)

    # Escape single quotes for PowerShell -replace RHS by doubling them
    ps_title = safe_title.replace("'", "''")
    ps_body = safe_body.replace("'", "''")

    # XML toast template — single-quoted here-string prevents PowerShell interpolation
    ps_script = f"""
$ErrorActionPreference = 'Stop'
[void][Windows.UI.Notifications.ToastNotificationManager, Windows.UI.Notifications, ContentType=WindowsRuntime]
$template = @'
<toast>
  <visual>
    <binding template="ToastGeneric">
      <text>TITLE_PLACEHOLDER</text>
      <text>BODY_PLACEHOLDER</text>
    </binding>
  </visual>
</toast>
'@
$template = $template -replace 'TITLE_PLACEHOLDER', '{ps_title}'
$template = $template -replace 'BODY_PLACEHOLDER', '{ps_body}'
$xml = [Windows.Data.Xml.Dom.XmlDocument]::new()
$xml.LoadXml($template)
$toast = [Windows.UI.Notifications.ToastNotification]::new($xml)
$notifier = [Windows.UI.Notifications.ToastNotificationManager]::CreateToastNotifier('AIHound')
$notifier.Show($toast)
""".strip()

    exe = shutil.which("powershell.exe") or shutil.which("powershell")
    if exe is None:
        return False
    result = subprocess.run(
        [exe, "-NoProfile", "-NonInteractive", "-Command", ps_script],
        capture_output=True,
        text=True,
        timeout=10,
    )
    if result.returncode != 0:
        logger.debug("PowerShell toast failed (rc=%d): %s", result.returncode, result.stderr)
        return False
    return True
